src/Arara/Process/IPC/Semaphore.php

```diff
@@ -0,0 +1,132 @@
+<?php
+
+namespace Arara\Process\IPC ;
+
+/**
+ * This class provides Interprocess communication using UNIX Semaphores
+ *
+ * Code written by Nikolay and copied into Arara/Process under the terms of
+ * the MIT license.
+ *
+ * Full credit to Nikolay for his excellent class
+ * 
+ * @license http://opensource.org/licenses/MIT
+ * @author Nikolay Bondarenko http://misterion.ru 
+ */
+class Semaphore
+{
+    /**
+     * @var resource
+     */
+    protected $mutex;
+
+    /**
+     * @var string
+     */
+    protected $file;
+
+    /**
+     * @var bool
+     */
+    protected $isAcquired;
+
+    /**
+     * Creator process pid
+     *
+     * @var int
+     */
+    protected $pid;
+
+    public function __construct($key = null)
+    {
+        $this->pid = getmypid();
+
+        $this->isAcquired = false;
+        if ($key) {
+            $semKey = $key;
+        } else {
+            $this->file = tempnam(sys_get_temp_dir(), 's');
+            $semKey = ftok($this->file, 'a');
+        }
+
+        $this->mutex = sem_get($semKey, 1); //auto_release = 1 by default
+        if (!$this->mutex) {
+            throw new \RuntimeException('Unable to create the semaphore');
+        }
+    }
+
+    public function __destruct()
+    {
+        if ($this->isAcquired()) {
+            $this->release();
+        }
+
+        if ($this->pid === getmypid()) {
+            $this->remove();
+        }
```

Comment:
I'm still new to the PR thing with github and the etiquette that goes with it.  Hope I haven't done the wrong thing.  While I haven't got the capacity to work through the issues with the IPC classes, I can re-issue the PR if you would like to work on it yourself?  Is that possible with PRs?  I haven't deleted the branch in my personal repo, so I should be able to re-open it again.

Would you like me to do that?